File: src/main/java/ru/arc/hooks/packetevents/PacketEventsHook.java
package ru.arc.hooks.packetevents;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;

import com.github.retrooper.packetevents.PacketEvents;
import com.github.retrooper.packetevents.protocol.entity.data.EntityData;
import com.github.retrooper.packetevents.protocol.entity.data.EntityDataTypes;
import com.github.retrooper.packetevents.protocol.entity.type.EntityTypes;
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerDestroyEntities;
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerEntityMetadata;
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerSpawnEntity;
import io.github.retrooper.packetevents.util.SpigotConversionUtil;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import ru.arc.ARC;

import static ru.arc.util.Logging.debug;

public class PacketEventsHook {


    public List<Integer> createDisplayBlocks(final List<BlockDisplayReq> requests, Player player) {
        if (requests.isEmpty()) {
            return List.of();
        }
        debug("Creating display blocks for player {}", player.getName());
        final List<Integer> entityIds = requests.stream()
                .map(request -> ThreadLocalRandom.current().nextInt(Integer.MIN_VALUE, Integer.MAX_VALUE))
                .collect(Collectors.toList());
        runOnMainThread(() -> {
            if (!player.isOnline()) {
                return;
            }
            int i = 0;
            for (BlockDisplayReq request : requests) {
                int entityId = entityIds.get(i++);
                WrapperPlayServerSpawnEntity spawnEntityPacket = new WrapperPlayServerSpawnEntity(
                        entityId,
                        UUID.randomUUID(),
                        EntityTypes.BLOCK_DISPLAY,
                        SpigotConversionUtil.fromBukkitLocation(request.location()),
                        0f,
                        0,
                        null
                );

                WrapperPlayServerEntityMetadata entityMetadataPacket = new WrapperPlayServerEntityMetadata(
                        entityId,
                        Collections.singletonList(
                                new EntityData(23, EntityDataTypes.BLOCK_STATE, SpigotConversionUtil.fromBukkitBlockData(request.data()).getGlobalId())
                        )
                );

                spawnEntityPacket.setEntityId(entityId);
                spawnEntityPacket.setUUID(Optional.of(UUID.randomUUID()));
                entityMetadataPacket.setEntityId(entityId);
                PacketEvents.getAPI().getPlayerManager().sendPacket(player, spawnEntityPacket);
                PacketEvents.getAPI().getPlayerManager().sendPacket(player, entityMetadataPacket);
            }
        });
        return entityIds;
    }

    public void removeDisplayBlocks(List<Integer> entityIds, Player player) {
        if (entityIds == null || entityIds.isEmpty()) {
            return;
        }
        debug("Removing {} display blocks for player {}", entityIds.size(), player.getName());
        runOnMainThread(() -> {
            if (!player.isOnline()) {
                return;
            }
            var packet = new WrapperPlayServerDestroyEntities(entityIds.stream().mapToInt(i -> i).toArray());
            PacketEvents.getAPI().getPlayerManager().sendPacket(player, packet);
        });
    }

    private void runOnMainThread(Runnable action) {
        if (Bukkit.isPrimaryThread()) {
            action.run();
        } else {
            Bukkit.getScheduler().runTask(ARC.getInstance(), action);
        }
    }

}

File: src/main/kotlin/ru/arc/hooks/packetevents/PacketEventsHook.kt
package ru.arc.hooks.packetevents

import com.github.retrooper.packetevents.PacketEvents
import com.github.retrooper.packetevents.protocol.entity.data.EntityData
import com.github.retrooper.packetevents.protocol.entity.data.EntityDataTypes
import com.github.retrooper.packetevents.protocol.entity.type.EntityTypes
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerDestroyEntities
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerEntityMetadata
import com.github.retrooper.packetevents.wrapper.play.server.WrapperPlayServerSpawnEntity
import io.github.retrooper.packetevents.util.SpigotConversionUtil
import net.kyori.adventure.util.TriState
import org.bukkit.Bukkit
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import ru.arc.core.sync
import ru.arc.util.Logging.debug
import java.util.Optional
import java.util.UUID
import java.util.concurrent.ThreadLocalRandom

class PacketEventsHook {

    /** Sends common entity flags to one viewer without changing server-side entity state. */
    fun setEntityInvisibleFor(entity: LivingEntity, player: Player, invisible: Boolean) {
        runOnMainThread {
            if (!player.isOnline || !entity.isValid) return@runOnMainThread
            val metadata =
                WrapperPlayServerEntityMetadata(
                    entity.entityId,
                    listOf(EntityData(0, EntityDataTypes.BYTE, commonEntityFlags(entity, invisibleForViewer = invisible))),
                )
            PacketEvents.getAPI().playerManager.sendPacket(player, metadata)
        }
    }

    /** Sends a viewer-only glow outline without changing the entity for other players. */
    fun setEntityGlowingFor(entity: LivingEntity, player: Player, glowing: Boolean) {
        runOnMainThread {
            if (!player.isOnline || !entity.isValid) return@runOnMainThread
            val metadata =
                WrapperPlayServerEntityMetadata(
                    entity.entityId,
                    listOf(EntityData(0, EntityDataTypes.BYTE, commonEntityFlags(entity, glowingForViewer = glowing))),
                )
            PacketEvents.getAPI().playerManager.sendPacket(player, metadata)
        }
    }

    fun createDisplayBlocks(requests: List<BlockDisplayReq>, player: Player): List<Int> {
        if (requests.isEmpty()) return emptyList()
        debug("Creating display blocks for player {}", player.name)

        val entityIds = requests.map { ThreadLocalRandom.current().nextInt(Int.MIN_VALUE, Int.MAX_VALUE) }

        runOnMainThread {
            if (!player.isOnline) return@runOnMainThread
            requests.forEachIndexed { i, request ->
                val entityId = entityIds[i]
                val spawnEntityPacket = WrapperPlayServerSpawnEntity(
                    entityId,
                    UUID.randomUUID(),
                    EntityTypes.BLOCK_DISPLAY,
                    SpigotConversionUtil.fromBukkitLocation(request.location),
                    0f, 0, null,
                )
                val entityMetadataPacket = WrapperPlayServerEntityMetadata(
                    entityId,
                    listOf(EntityData(23, EntityDataTypes.BLOCK_STATE, SpigotConversionUtil.fromBukkitBlockData(request.data).globalId)),
                )
                spawnEntityPacket.entityId = entityId
                spawnEntityPacket.uuid = Optional.of(UUID.randomUUID())
                entityMetadataPacket.entityId = entityId
                PacketEvents.getAPI().playerManager.sendPacket(player, spawnEntityPacket)
                PacketEvents.getAPI().playerManager.sendPacket(player, entityMetadataPacket)
            }
        }
        return entityIds
    }

    fun removeDisplayBlocks(entityIds: List<Int>?, player: Player) {
        if (entityIds.isNullOrEmpty()) return
        debug("Removing {} display blocks for player {}", entityIds.size, player.name)
        runOnMainThread {
            if (!player.isOnline) return@runOnMainThread
            val packet = WrapperPlayServerDestroyEntities(*entityIds.toIntArray())
            PacketEvents.getAPI().playerManager.sendPacket(player, packet)
        }
    }

    private fun runOnMainThread(action: () -> Unit) {
        if (Bukkit.isPrimaryThread()) action()
        else sync(action)
    }
}

internal fun commonEntityFlags(
    entity: LivingEntity,
    invisibleForViewer: Boolean? = null,
    glowingForViewer: Boolean? = null,
): Byte {
    var flags = 0
    val invisible = invisibleForViewer ?: entity.isInvisible
    val glowing = glowingForViewer ?: entity.isGlowing
    if (entity.fireTicks > 0 || entity.visualFire == TriState.TRUE) flags = flags or 0x01
    if (entity.isSneaking) flags = flags or 0x02
    if (entity.isSwimming) flags = flags or 0x10
    if (invisible) flags = flags or 0x20
    if (!invisible && glowing) flags = flags or 0x40
    if (entity.isGliding) flags = flags or 0x80
    return flags.toByte()
}
